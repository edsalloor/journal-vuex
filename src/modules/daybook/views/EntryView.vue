<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
      </div>

      <div>
        <button
          v-if="entry.id"
          class="btn btn-danger mx-2"
          @click="onDeleteEntry"
        >
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>
        
        <button
          class="btn btn-primary"
          @click="onUploadImage"
        >
          Upload Photo
          <i class="fa fa-upload"></i>
          <input
            @change="onSelectedImage"
            accept="image/png, image/jpeg, image/webp"
            ref="imageSelector"
            type="file"
            v-show="false"
          >
        </button>
      </div>
    </div>

    <hr>

    <div class="d-flex flex-column px-3 h-75">
      <textarea
        v-model="entry.text"
        placeholder="What's happened today?"
      ></textarea>
    </div>

    <img
      v-if="localImage || entry.picture"
      :src="localImage || entry.picture"
      alt="entry-picture"
      class="img-thumbnail"
    >
  </template>

  <Fab
    icon="fa-save"
    @fab:click="saveEntry"
  />
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapActions, mapGetters } from 'vuex';
import Swal from 'sweetalert2';

import getDayMonthYear from '@/modules/daybook/helpers/getDayMonthYear';
import uploadImage from '@/modules/daybook/helpers/uploadImage';

export default {
  name: 'EntryView',
  props: {
    id: {
      required: true,
      type: String
    }
  },
  components: {
    Fab: defineAsyncComponent(() => import('@/modules/daybook/components/Fab.vue'))
  },
  data () {
    return {
      entry: null,
      localImage: null,
      imageFile: null
    };
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    }
  },
  methods: {
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
    loadEntry() {
      let entry;

      if (this.id === 'new') {
        entry = {
          text: '', date: new Date().getTime()
        };
      } else {
        entry = this.getEntryById(this.id);
        if(!entry) return this.$router.push({ name: 'no-entry' });
      }

      this.entry = entry;
      this.localImage = null;
    },
    async saveEntry() {
      Swal.fire({
        title: 'Wait please...',
        allowOutsideClick: false
      });
      Swal.showLoading();

      const imageUrl = await uploadImage(this.imageFile);
      this.entry.picture = imageUrl;

      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const { id } = await this.createEntry(this.entry);
        this.$router.push({name: 'entry', params: { id }});
      }

      this.localImage = null;
      Swal.fire('Saved', 'Entry successfully created.', 'success');
    },
    async onDeleteEntry() {
      const { isConfirmed } = await Swal.fire({
        title: 'Do you want to delete this entry?',
        text: 'This action can not be undone.',
        showDenyButton: true,
        confirmButtonText: 'Delete'
      });

      if (!isConfirmed) return;

      Swal.fire({
        title: 'Please wait...',
        allowOutsideClick: false
      });
      Swal.showLoading();

      await this.deleteEntry(this.entry.id);
      this.$router.push({ name: 'no-entry' });

      Swal.fire('Entry deleted', '', 'success');
    },
    onSelectedImage(event) {
      const file = event.target.files[0];

      if (!file) {
        this.localImage = null;
        this.imageFile = null;
        return;
      }

      this.imageFile = file;

      const fr = new FileReader();
      fr.onload = () => this.localImage = fr.result;
      fr.readAsDataURL(file);
    },
    onUploadImage() {
      this.$refs.imageSelector.click();
    }
  },
  created() {
    this.loadEntry();
  },
  watch: {
    id() {
      this.loadEntry();
    }
  }
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  bottom: 150px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
  position: fixed;
  right: 20px;
  width: 200px;
}
</style>
