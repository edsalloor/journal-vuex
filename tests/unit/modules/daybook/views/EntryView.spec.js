import { shallowMount } from '@vue/test-utils';
import Swal from 'sweetalert2';
import { createStore } from 'vuex';

import defaultInitialState from '@/../tests/unit/mocks/journalState';
import journalModule from '@/modules/daybook/store/journal';
import EntryView from '@/modules/daybook/views/EntryView';

jest.mock('sweetalert2', () => ({
  close: jest.fn(),
  fire: jest.fn(),
  showLoading: jest.fn()
}));

const setupStore = (initialState=defaultInitialState) => createStore({
  modules: {
    journal: {
      ...journalModule,
      state: { ...initialState }
    }
  }
});
const store = setupStore();
store.dispatch = jest.fn();

const routerMock = {
  push: jest.fn()
};

const setup = props => shallowMount(EntryView, {
  global: {
    mocks: {
      $router: routerMock
    },
    plugins: [ store ],
  },
  props: {
    id: defaultInitialState.entries[0].id,
    ...props
  }
});

describe('<EntryView />', () => {
  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should redirect to "/no-entry" endpoint when entry id does not exist', () => {
    setup({ id: 'non-existing-id' });

    expect(routerMock.push).toHaveBeenCalledTimes(1);
    expect(routerMock.push).toHaveBeenCalledWith({name: 'no-entry'});
  });

  it('should render entry when entry id exists', () => {
    wrapper = setup();

    expect(wrapper.html()).toMatchSnapshot();
    expect(routerMock.push).not.toHaveBeenCalled();
  });

  it('should delete entry and redirect to "/no-entry" endpoint', (done) => {
    wrapper = setup();
    Swal.fire.mockReturnValue(Promise.resolve({ isConfirmed: true }));

    wrapper.find('.btn-danger').trigger('click');

    setTimeout(() => {
      expect(Swal.fire).toHaveBeenCalledTimes(3);
      expect(Swal.fire).toHaveBeenNthCalledWith(1, {
        title: 'Do you want to delete this entry?',
        text: 'This action can not be undone.',
        showDenyButton: true,
        confirmButtonText: 'Delete'
      });
      expect(Swal.fire).toHaveBeenNthCalledWith(2, {
        title: 'Please wait...',
        allowOutsideClick: false
      });
      expect(Swal.fire).toHaveBeenNthCalledWith(3, 'Entry deleted', '', 'success');

      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(
        'journal/deleteEntry',
        defaultInitialState.entries[0].id
      );

      expect(routerMock.push).toHaveBeenCalledTimes(1);
      expect(routerMock.push).toHaveBeenCalledWith({name: 'no-entry'});
      done();
    }, 1);
  });
});
