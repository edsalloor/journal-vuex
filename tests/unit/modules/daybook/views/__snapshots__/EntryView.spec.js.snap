// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntryView /> should render entry when entry id exists 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">23</span><span class="mx-1 fs-3">July</span><span class="mx-2 fs-4 fw-light">2021, Friday</span></div>
  <div><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload Photo <i class="fa fa-upload"></i><input accept="image/png, image/jpeg, image/webp" type="file" style="display: none;"></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What's happened today?"></textarea></div>
<img src="https://res.cloudinary.com/sample.jpg" alt="entry-picture" class="img-thumbnail">
<fab-stub icon="fa-save"></fab-stub>
`;
